### DISCUSSION TITLE
Sitemap module : how to write file in .output/public at build time ?

### DISCUSSION THREAD

**Discussion by benoitdemaegdt** [nuxt/nuxt#17691] [Questions] [answered]
Hey there 👋 

I'm trying to fix this minimalistic sitemap module for nuxt3 : https://github.com/benoitdemaegdt/nuxt3-sitemap

It does not work since `nitro:generate` hook was removed 2 weeks ago.

So far, it is able to : 
- get all static routes
- generate a sitemap.xml file inside the project

But I cannot figure out how to write this sitemap.xml file inside `.output/public` directory at build time. I guess I also need to tell nitro that this file can be served when calling `<YOUR_DOMAIN>/sitemap.xml` ?

Any idea ? 💡 

This is my latest code here 👇 
```js
import { mkdirSync, writeFileSync } from 'fs'
import { Readable } from 'stream'
import { dirname } from 'path'
import { SitemapStream, streamToPromise } from 'sitemap'
import { defineNuxtModule, createResolver } from '@nuxt/kit'

export default defineNuxtModule({
  meta: {
    name: 'sitemap',
    version: '0.0.1',
    configKey: 'sitemap',
    compatibility: { nuxt: '^3.0.0' },
  },
  defaults: {
    hostname: 'http://localhost:3000',
  },
  async setup(options, nuxt) {
    async function generateSitemap(routes) {
      const sitemapRoutes = routes.map(route => route.path)

      // https://github.com/ekalinin/sitemap.js#generate-a-one-time-sitemap-from-a-list-of-urls
      const stream = new SitemapStream({ hostname: options.hostname })
      return streamToPromise(Readable.from(sitemapRoutes).pipe(stream)).then(data =>
        data.toString()
      )
    }

    function createSitemapFile(sitemap, filepath) {
      const dirPath = dirname(filepath)
      mkdirSync(dirPath, { recursive: true })
      writeFileSync(filepath, sitemap)
    }

    nuxt.hook('pages:extend', async pages => {
      const routes = pages

      const resolver = createResolver(import.meta.url)
      const filePath = await resolver.resolvePath('~/.sitemap/sitemap.xml')
      const sitemap = await generateSitemap(routes)
      createSitemapFile(sitemap, filePath)

      // TODO: write (or copy ?) sitemap file inside public folder
      // this is how partytown module interacts with public folder, but it does not work here
      // nuxt.options.nitro.publicAssets = nuxt.options.nitro.publicAssets || []
      // nuxt.options.nitro.publicAssets.push({
      //   baseURL: '/sitemap.xml',
      //   dir: dirname(filePath),
      // })
    })
  },
})
```

**Comment by Diizzayy** [comment, marked as answer]
> But I cannot figure out how to write this sitemap.xml file inside .output/public directory at build time

@benoitdemaegdt You're on the right track, though a few points to take note of.

- your publicAsset entry needs to be added outside of the nuxt hook
- the publicAsset `baseURL` is meant to be the relative path that the asset will be accessed within. Upon build, nitro ensures that the publicAsset `dir` exists and is indeed a directory, if so it then copies all the files within the given directory to `public/<baseURL>`, so for a sitemap, simple passing `/` as the baseURL will allow it to be accessible at `example.com/sitemap.xml`

The code below now works as you'd expect.

I also changed the sitemap output to a temp dir, just to avoid the need for it to be added to gitignore

```ts
import { mkdirSync, writeFileSync } from 'fs'
import { Readable } from 'stream'
import { dirname } from 'path'
import { SitemapStream, streamToPromise } from 'sitemap'
import { defineNuxtModule, createResolver } from '@nuxt/kit'

export default defineNuxtModule({
  meta: {
    name: 'sitemap',
    version: '0.0.1',
    configKey: 'sitemap',
    compatibility: { nuxt: '^3.0.0' },
  },
  defaults: {
    hostname: 'http://localhost:3000',
  },
  async setup(options, nuxt) {
    async function generateSitemap(routes) {
      const sitemapRoutes = routes.map(route => route.path)

      // https://github.com/ekalinin/sitemap.js#generate-a-one-time-sitemap-from-a-list-of-urls
      const stream = new SitemapStream({ hostname: options.hostname })
      return streamToPromise(Readable.from(sitemapRoutes).pipe(stream)).then(data =>
        data.toString()
      )
    }

    function createSitemapFile(sitemap, filepath) {
      const dirPath = dirname(filepath)
      mkdirSync(dirPath, { recursive: true })
      writeFileSync(filepath, sitemap)
    }

    const resolver = createResolver(import.meta.url)
    const filePath = resolver.resolve(nuxt.options.srcDir, 'node_modules/.cache/.sitemap/sitemap.xml')

    nuxt.options.nitro.publicAssets = nuxt.options.nitro.publicAssets || []
    nuxt.options.nitro.publicAssets.push({
      baseURL: '/',
      dir: dirname(filePath),
    })

    nuxt.hook('pages:extend', async (pages) => {
      const sitemap = await generateSitemap(pages)
      createSitemapFile(sitemap, filePath)
    })
  }
})
```

**Reply by benoitdemaegdt** [reply]
Thanks a lot @Diizzayy for this answer !

This code is perfectly working, and I now have a better understanding of how all of this work 😃

**Reply by florian-lefebvre** [reply]
Great to see the improvements compared to the original version! Would be great to get routes dynamically

**Reply by benoitdemaegdt** [reply]
hey @florian-lefebvre  👋 

I had a quick look at the official nuxt/sitemap module.
It seems that it handle dynamic routes by adding a `routes` option 👉  https://sitemap.nuxtjs.org/usage/sitemap

So basically, what is missing here for handling dynamic routes is to add this `routes` option (can be an array or a function).

It might look like this (not clean and not tested, but you get the idea) : 

```
   ...
    nuxt.hook('pages:extend', async (staticRoutes) => {
     const dynamicRoutes = Array.isArray(options.routes) ? options.routes : await options.routes()
      const sitemap = await generateSitemap([ ...staticRoutes, ...dynamicRoutes ])
      createSitemapFile(sitemap, filePath)
    })
```

Then in nuxt.config, it would be the developer responsibility to fetch and properly format dynamic routes.

Example using nuxt2 & nuxt/content in one of my project : 
- sitemap module config : https://github.com/benoitdemaegdt/monpetitsommet/blob/master/nuxt.config.js#L100
- get all dynamic routes : https://github.com/benoitdemaegdt/monpetitsommet/blob/master/nuxt.config.js#L123

In order to get a better understanding of the use case, may I ask how you intent to use it ? What are you building and with which tools ? I guess you are using a CMS ?

**Reply by florian-lefebvre** [reply]
Just a list of static elements but using a module options property as a function was broking everything. In fact I'd like the same as in https://github.com/harlan-zw/unlighthouse, where all routes are fetched automatically. I didn't manage to understand how it is done but I think it's the best way to generate the sitemap.
